src/latency.c

```diff
@@ -490,6 +490,83 @@ sds createLatencyReport(void) {
 
+/* latencyCommand() helper to produce an map of time buckets,
+ * each representing a latency range,
+ * between 1 nanosecond and roughly 1 second.
+ * Each bucket covers twice the previous bucket's range.
+ * Empty buckets are not printed.
+ * Everything above 1 sec is considered +Inf.
+ * At max there will be log2(1000000000)=30 buckets */
+void fillCommandCDF(client *c, struct hdr_histogram* histogram){
+    addReplyMapLen(c,2);
+    addReplyBulkCString(c,"calls");
+    addReplyLongLong(c,(long long) histogram->total_count);
+    addReplyBulkCString(c,"histogram_usec");
+    void *replylen = addReplyDeferredLen(c);
+    int samples = 0;
+    struct hdr_iter iter;
+    hdr_iter_log_init(&iter,histogram,1024,2);
+    int64_t previous_count = 0;
+    while (hdr_iter_next(&iter)) {
+        const int64_t micros = iter.highest_equivalent_value / 1000;
+        const int64_t cumulative_count = iter.cumulative_count;
+        if(cumulative_count > previous_count){
+            addReplyLongLong(c,(long long) micros);
+            addReplyLongLong(c,(long long) cumulative_count);
+            samples++;
+        }
+        previous_count = cumulative_count;
+    }
+    setDeferredMapLen(c,replylen,samples);
+}
+
+/* latencyCommand() helper to produce for all commands,
+ * a per command cumulative distribution of latencies. */
+void latencyAllCommandsFillCDF(client *c) {
+    dictIterator *di = dictGetSafeIterator(server.commands);
+    dictEntry *de;
+    struct redisCommand *cmd;
+    void *replylen = addReplyDeferredLen(c);
+    int command_with_data = 0;
+    while((de = dictNext(di)) != NULL) {
+        char *tmpsafe;
+        cmd = (struct redisCommand *) dictGetVal(de);
+        if (!cmd->latency_histogram)
+            continue;
+        const char* cmdname = getSafeInfoString(cmd->name, strlen(cmd->name), &tmpsafe);
+        addReplyBulkCString(c,cmdname);
+        fillCommandCDF(c, cmd->latency_histogram);
+        if (tmpsafe != NULL) zfree(tmpsafe);
+        command_with_data++;
+    }
+    dictReleaseIterator(di);
+    setDeferredMapLen(c,replylen,command_with_data);
+}
+
+/* latencyCommand() helper to produce for a specific command set,
+ * a per command cumulative distribution of latencies. */
+void latencySpecificCommandsFillCDF(client *c) {
+    addReplyMapLen(c,c->argc - 2);
+    for (int j = 2; j < c->argc; j++){
+        struct redisCommand *cmd = dictFetchValue(server.commands, c->argv[j]->ptr);
+        char *tmpsafe;
+        const char* cmdname = getSafeInfoString(c->argv[j]->ptr, strlen(c->argv[j]->ptr), &tmpsafe);
+        addReplyBulkCString(c,cmdname);
+        // If the command does not exist we reply with an empty map
+        if (cmd == NULL) {
+            addReplyMapLen(c,0);
+            continue;
+        }
+        // If no latency info we reply with an empty map
+        if (!cmd->latency_histogram) {
+            addReplyMapLen(c,0);
+            continue;
+        }
```

Comment:
i'm guessing we should skip these too.
the only reason i can think of for replying with an empty map is to distinguish these from unknown commands, but with your current code they're the same, so i guess that's not the concern, but the ordering of the reply.

another possible concern is that clients will be able to more easily parse this the same as non-empty ones, but for that we need to create the nested map, with the `calls` field set to 0.

Reply:
> another possible concern is that clients will be able to more easily parse this the same as non-empty ones, but for that we need to create the nested map, with the calls field set to 0.

@oranagra I've changed it to reply with the same "schema" but the calls set to 0 and with an empty histogram map.

Reply:
i think we can just skip them.. why bother to add so much details (two levels of nesting) for en empty response?
it's unordered anyway, and we skip empty ones in the "all" variant.